### DISCUSSION TITLE
Amend commit time requires kart upgrade?

### DISCUSSION THREAD

**Discussion by alpha-beta-soup** [koordinates/kart#916] [Q&A] [answered]
Is it possible to amend the commit time? I'm interested in adapting some old data that is available as a series of point-in-time archives into a single Kart repository, but ideally I want the commit date to be the date the archive was made, not the datetime I've made the Kart commit. To me it seems sensible to make a commit, then amend the commit time. Starting with a new repo and GPKG with only an init commit:

```shell
kart git commit --amend --date="Tue 28 Mar 03:01:20 2023 +0000" --no-edit
```
```
[main ef95198] Import from data.gpkg:layer to layer/
 Date: Tue Mar 28 03:01:20 2023 +0000
```

So far so good.

```shell
kart log
```
```
Error: This Kart repo uses Datasets v1, but Kart 0.14.1 only supports Datasets v2 or v3.
Use `kart upgrade SOURCE DEST` to upgrade this repo to the supported version.
```

Kart log had no issues before; i.e. I don't get this error before the amend operation. Perhaps Kart version checking relies on commit dates?

If I `kart upgrade` as suggested then everything's fine, but the upgrade step seems a little odd. Is there another way to do this that doesn't require the intermediate `kart upgrade`?

---

`kart --version`:

```
Kart v0.14.1, Copyright (c) Kart Contributors
» GDAL v3.6.3; PROJ v9.2.0; PDAL v2.5.3
» PyGit2 v1.12.1; Libgit2 v1.6.4; Git v2.38.1; Git LFS v3.3.0
» SQLAlchemy v1.4.45; pysqlite3 v2.6.0/v3.40.1; SpatiaLite v5.0.1; Libpq v15.0.3
Executed via helper, SID=1540402 PID=1550215
```

---

~~If I `kart upgrade` as suggested then everything's fine~~

Actually my upgraded repository has the correct commit message and amended date, but no working copy. If I create one, it's empty. I'm new to kart so might be missing something though. When I upgrade, how do I also keep the working copy?

**Comment by olsen232** [comment, marked as answer]
OK you've found some rough edges of Kart here unfortunately - I'll have to file some issues against myself.

Things have gone wrong earlier than the `kart upgrade` - this is a red herring, the `kart git` command you ran has corrupted your commit already and left it empty. (Running arbitrary `kart git` commands is not guaranteed to leave your kart repository in a good state, since, Git doesn't understand what a Kart repository is or how to make sure it is left in a good state. That needs to be better documented, or the `kart git` command should be better hidden - that's the first issue)

Since Git doesn't understand the Kart working copy, for technical reasons, the command you ran actually overwrites your commit with an empty commit, which is not what you want. This is why I should hide this command better.
Your commit now contains nothing, and in particular doesn't contain a marker that marks it as a Kart repo. But, since it looks like a Kart repo in other ways, Kart assumes it may be a Kart repo from the earliest version of Kart. This is the second issue - the version detection logic and error message should be improved, since, following the `kart upgrade` advice didn't fix anything at all - the repo remained empty.

Finally, the sort of thing you are trying to do is not so straight-forward in Kart - Git has innumerable commands, flags, environment variables, etc that let you modify your repository in any way, Kart has much less, and using the Git ones can go wrong. Nevertheless, it is possible - here is a Kart command which does it:
`GIT_COMMITTER_DATE="Tue 28 Mar 03:01:20 2023 +0000" kart commit-files -m "Import from data.gpkg:layer to layer/" --amend`

(This command generally lets you store arbitrary files in your Kart repo, but in this case we're telling it to add zero files but to amend the commit and keep the same message)

But first you'll need to start over to fix the previous command - there are other ways, but probably recreating the repo from scratch is easiest.

**Reply by alpha-beta-soup** [reply]
Thanks. I'm happy starting over, since the present "master" isn't a Kart repo at all and I don't plan to change that. This is experimental for me.

I did think I was hitting a square peg in a round hole, but I really like this idea of marking change over time, particularly when the commit times are accurate. It seems that it's at least possible, which is honestly great. Thanks a lot for your effort.

**Comment by alpha-beta-soup** [comment]
Just as a comment for anyone else who may be following this thread in the future, there's a difference between "AuthorDate" and "CommitDate" in git. It seems appropriate to adjust the CommitDate but to leave the AuthorDate as is. You can see both with `git log --pretty=fuller`.

**Reply by olsen232** [reply]
Good point. It's not obvious that there are two different dates since Kart / Git are a bit reluctant to show you them both for whatever reason - but you can do it with `git log --format=fuller`, which works perfectly well for Kart repos. Kart log doesn't support this option.

(Note: `git <command>` is mostly safe in a Kart repo since it won't touch the working copy - if it tries to it instead fails with a message about an unrecognised kart extension. Of course, if you know the full repertoire of git commands, you can eventually corrupt anything. `kart git <command>` provides git with an empty working copy that it understands, which makes it slightly more powerful but a lot more dangerous - as noted above, it should be treated as part of Kart's internals.)

**Comment by alpha-beta-soup** [comment]
Link to my experimentation https://gist.github.com/alpha-beta-soup/26d7116bbc53a47fa3ce9cf7f9cd76f2
